internal/cluster/cluster_service.go

```diff
@@ -0,0 +1,329 @@
+		if atomic.CompareAndSwapInt32(&s.startChAtom, 0, 1) {
+			close(s.startCh)
+			s.startCh = nil
+		}
+		if len(added) > 0 {
+			s.eventDispatcher.Publish(NewMembersAdded(added))
+		}
+		if len(removed) > 0 {
+			s.eventDispatcher.Publish(NewMemberRemoved(removed))
+		}
+	}
+}
+
+func (s *ServiceImpl) sendMemberListViewRequest(conn *Connection) {
+	request := codec.EncodeClientAddClusterViewListenerRequest()
+	inv := s.invocationFactory.NewConnectionBoundInvocation(request, -1, nil, conn, s.config.InvocationTimeout, func(response *proto.ClientMessage) {
+		codec.HandleClientAddClusterViewListener(response, func(version int32, memberInfos []pubcluster.MemberInfo) {
+			s.logger.Debug(func() string { return "members updated" })
+			s.eventDispatcher.Publish(NewMembersUpdated(memberInfos, version))
+		}, func(version int32, partitions []proto.Pair) {
+			s.eventDispatcher.Publish(NewPartitionsUpdated(partitions, version))
+			s.logger.Debug(func() string { return "partitions updated" })
+		})
+	})
+	s.requestCh <- inv
+	if _, err := inv.Get(); err != nil {
+		s.logger.Error(err)
+	}
+}
+
+func (s *ServiceImpl) listenPartitionsLoaded() {
+	subscriptionID := event.MakeSubscriptionID(s.enableSmartRouting)
+	handler := func(event event.Event) {
+		s.enableSmartRouting()
+		s.eventDispatcher.Unsubscribe(EventPartitionsLoaded, subscriptionID)
+	}
+	s.eventDispatcher.Subscribe(EventPartitionsLoaded, subscriptionID, handler)
+}
+
+func (s *ServiceImpl) enableSmartRouting() {
+	s.logger.Infof("enabling smart routing")
+	atomic.StoreInt32(&s.smartRouting, smartRoutingEnabled)
+}
+
+func (m *ServiceImpl) logStatus() {
+	ticker := time.NewTicker(10 * time.Second)
+	for {
+		select {
+		case <-m.doneCh:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			m.membersMap.Info(func(members map[string]*Member) {
+				m.logger.Trace(func() string {
+					mems := map[string]string{}
+					for uuid, member := range members {
+						mems[uuid] = member.Address().String()
+					}
+					return fmt.Sprintf("members: %#v", mems)
+				})
+			})
+		}
+	}
+}
+
+type AddrSet struct {
+	addrs map[string]*pubcluster.AddressImpl
+}
+
+func NewAddrSet() AddrSet {
+	return AddrSet{addrs: map[string]*pubcluster.AddressImpl{}}
+}
+
+func (a AddrSet) AddAddr(addr *pubcluster.AddressImpl) {
+	a.addrs[addr.String()] = addr
+}
+
+func (a AddrSet) AddAddrs(addrs []*pubcluster.AddressImpl) {
+	for _, addr := range addrs {
+		a.AddAddr(addr)
+	}
+}
+
+func (a AddrSet) Addrs() []*pubcluster.AddressImpl {
+	addrs := make([]*pubcluster.AddressImpl, 0, len(a.addrs))
+	for _, addr := range a.addrs {
+		addrs = append(addrs, addr)
+	}
+	return addrs
+}
+
+type membersMap struct {
+	members          map[string]*Member
+	addrToMemberUUID map[string]string
+	membersMu        *sync.RWMutex
+	version          int32
+}
+
+func newMembersMap() membersMap {
+	return membersMap{
+		members:          map[string]*Member{},
+		addrToMemberUUID: map[string]string{},
+		membersMu:        &sync.RWMutex{},
```

Comment:
Mutexes should not be copied. Making it a value type here may cause it to be copied. https://eli.thegreenplace.net/2018/beware-of-copying-mutexes-in-go/

Reply:
It's simple to avoid copying mutexes. Values for mutexes is a common approach and it can be seen in many popular open source projects written in Go and personally I don't see any problems with it.

Reply:
I would argue that making the mutex a pointer and initializing it is easier than remembering to not copy it. I'll mark this as resolved, but please feel free to reopen in case we should discuss that further.

Reply:
No need to reopen since this comment was a `nit` (`n`ot `i`mportant `t`hought), hence it's subjective and optional by the definition.